camera/camera_manager.cc

```diff
@@ -16,68 +50,41 @@ CameraManager::CameraManager() {
-    if (filename == "." || filename == ".."){ 
+    if (filename == "." || filename == "..") {
       continue;
     }
-    std::optional<YAML::Node> config = read_YAML(filename);
-    std::optional<Camera> camera = parse_config(*config);
+
+    const std::optional<YAML::Node> config = read_YAML(filename);
+    if (!config) {
+      throw std::runtime_error("Could not open config");
+    }
+
+    const std::optional<Camera> camera = extract_v4l_path(*config);
     if (camera.has_value()) {
       camera_map_[camera->serial_number] = *camera;
     }
-  }
-}
 
-std::optional<YAML::Node> CameraManager::read_YAML(const std::string& filepath) const {
-  try {
-    return YAML::LoadFile(filepath);
-  } catch (YAML::BadFile e) {
-    const std::string err = std::string("Could not find YAML file ") + filepath;
-    throw std::runtime_error(err);
+    const auto calibration = extract_calibration(*config);
+    calibration_map_[camera->serial_number] = calibration;
   }
 }
 
-std::optional<Camera> CameraManager::parse_config(const YAML::Node& cfg) const {
-  Camera camera;
-  // path & video index
-  camera.serial_number = cfg["serial_number"].as<std::string>();
-  camera.v4l_path = "/dev/v4l/by-id/" + cfg["v4l_path"].as<std::string>();
-  const auto optional_camera_index = follow_v4l_path(camera.v4l_path);
-  
-  if (!optional_camera_index.has_value()){
-    return std::nullopt;
-  }
-  else{
-    camera.video_index = optional_camera_index.value();
+Calibration CameraManager::get_calibration(const std::string& serial_number) const {
+  if (calibration_map_.count(serial_number) == 0) {
+    const std::string err = std::string("Camera serial number is not known.");
+    throw std::runtime_error(err);
   }
-  // calibration values
-  std::vector<double> camera_matrix_values = cfg["camera_matrix"].as<std::vector<double>>();
-  camera.calibration.camera_matrix = cv::Mat(3, 3, CV_64F, camera_matrix_values.data()).clone();
-  std::vector<double> distortion_coefficient_values = cfg["distortion_coefficients"].as<std::vector<double>>();
-  camera.calibration.distortion_coefficients = cv::Mat(1, 5, CV_64F, distortion_coefficient_values.data()).clone();
-  return camera;
-}
-
-std::optional<int> CameraManager::follow_v4l_path(const std::string& path) const {
-  std::error_code ec;
-  const std::string v4l_path = fs::canonical(path, ec).string();
-  if (ec) {
-    return std::nullopt;
-  } 
-  // note that this will break for >10 cameras
-  // TODO better way
-  const int index = v4l_path.back() - '0';
-  return index;
+  return calibration_map_.at(serial_number);
 }
 
 Camera CameraManager::get_camera(const std::string& serial_number) const {
   if (camera_map_.count(serial_number) == 0) {
-    const std::string err = std::string(
-        "The camera config passed to start the camera BQ specified "
-        "a serial number that doesn't match the camera we have plugged in.");
+    const std::string err = std::string("Camera is not plugged in");
```

Comment:
Change to eg "The camera we are looking for is not plugged in" to differentiate between being connected to no cameras and being connected to the wrong camera.  I think the old error message covered this.  Struggling to get cameras connected in the presence of various issues such as opening the container before plugging in the camera  means that rich error messages will have lots of value.